tornado/platform/asyncio.py

```diff
@@ -244,10 +237,14 @@ def remove_timeout(self, timeout: object) -> None:
         timeout.cancel()  # type: ignore
 
     def add_callback(self, callback: Callable, *args: Any, **kwargs: Any) -> None:
-        if threading.get_ident() == self._thread_identity:
-            call_soon = self.asyncio_loop.call_soon
-        else:
+        try:
+            if asyncio.get_running_loop() is self.asyncio_loop:
+                call_soon = self.asyncio_loop.call_soon
+            else:
+                call_soon = self.asyncio_loop.call_soon_threadsafe
+        except RuntimeError:
             call_soon = self.asyncio_loop.call_soon_threadsafe
+
         try:
             call_soon(self._run_callback, functools.partial(callback, *args, **kwargs))
         except RuntimeError:
```

Comment:
```suggestion
        except RuntimeError:
            if self.asyncio_loop.get_debug():
                raise  # loop operating in debug mode uses are probably interested in exceptions
```

Reply:
Maybe. I take the opposite view of asyncio on the types of exceptions that can be raised here. When add_callback races with a closing event loop, everything's shutting down and any exceptions are just going to spam the logs in unhelpful ways, so I'd make that exception debug-only. The "thread-unsafe operation on wrong thread" exception, on the other hand, could result in much more subtle problems and is worth raising even outside of debug mode (or, if you're going to go to the expense of checking the thread, you might as well just do the right thing instead of having two separate methods). 

I wish asyncio had well-defined exceptions so we didn't have to do a blanket catch of RuntimeError (and AttributeError!) here. I think when I wrote this the closed loop error was the only one that could be raised. Now there are more (at least in debug mode), including checks on the type of callable. So maybe it is worth allowing exceptions to pass through in debug mode.